File: JavaScript/823-Binary Trees With Factors.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var numFactoredBinaryTrees = function(arr) {
    arr.sort(function(a,b){
        return a-b;
    });
    let hmap = new Map();
    let res = 0;
    for (let i=0; i<arr.length; i++){
        hmap.set(arr[i], 1);
        for (let j=0; j<i; j++){
            if (arr[i]%arr[j] === 0 && hmap.has(arr[i]/arr[j])){
                hmap.set(arr[i], hmap.get(arr[i]) + hmap.get(arr[j])*hmap.get(arr[i]/arr[j]) );
            }
        }
        res += hmap.get(arr[i])%(10**9 + 7);
    }
    return res%(10**9+7);
};

File: C++/823-Binary Trees With Factors.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        sort(arr.begin(), arr.end());
        unordered_map<int, long> dp;
        int kMod = 1e9+7;
        long res = 0;
        
        for (int i=0; i<arr.size(); i++){
            if (dp.count(arr[i])) continue;
            dp[arr[i]] = 1;
            for (int j=0; j<i; j++){
                if (arr[i]%arr[j] == 0 && dp.count(arr[i]/arr[j])){
                    dp[arr[i]] = (dp[arr[i]] + dp[arr[j]]*dp[arr[i]/arr[j]])%kMod;
                }
            }
            res = (res + dp[arr[i]])%kMod;
        }
        return res;
    }
};